---
layout: default
title: Lesson 7 - Differential methylation
parent: 3. Tutorial
nav_order: 7
description: A comprehensive guide to understanding epigenetics.
published: true
---

Final version
{: .label .label-green }

{: .important-title }
> Aim
>
> Perform a differential methylation analysis using methylKit

<!--
<br>
<details open markdown="block">
    <summary>
        <strong>Table of contents</strong>
    </summary>
    {: .text-delta }
- TOC
{:toc}
</details>
<br>
-->

MethylKit is an R package for the analysis of DNA methylation data. It can be used to analyze data obtained by high-througput bisulfite sequencing (BS-seq) or reduced representation bisulfite sequencing (RRBS). 

[Learn more about methylKit](https://www.bioconductor.org/packages/release/bioc/vignettes/methylKit/inst/doc/methylKit.html)


The methylation call data has a slightly different format compared to the Bismark output

```
 chr  start    end strand coverage numCs numTs
chr1  18001  19000      *       48    47     1
chr1  29001  30000      *      625     2   623
chr1  61001  62000      *       38     0    38
chr1  91001  92000      *       44     0    44
chr1 139001 140000      *       12    12     0
chr1 144001 145000      *       47    45     2
```

Thus if we want to read the data generated by Bismark, we will need to modify the order of the columns (for example using `awk`).


IN this tutorial we will use already formatted files for four different samples. 


### Activate the conda environment
{: .no_toc }

```bash
conda activate methylkit

# Move to the working directory if we want to save files 
cd /data2/student_space/st24_16_folder/epigenomics

# Open R
R

# Load the library
library(methylKit)
```


```r
# Create a list object with the path to the file of interest
file.list = list("/data2/biotecnologie_molecolari_magris/epigenomics/methylkit/dataset/301.rrbs.methylkit",
"/data2/biotecnologie_molecolari_magris/epigenomics/methylkit/dataset/302.rrbs.methylkit",
"/data2/biotecnologie_molecolari_magris/epigenomics/methylkit/dataset/310.rrbs.methylkit",
"/data2/biotecnologie_molecolari_magris/epigenomics/methylkit/dataset/311.rrbs.methylkit")
```


Load the files into methylKit objects. By defining the treatment function, we classify the samples in case or treated (1) vs controls (0). Samples 301 and 302 are controls, while samples 310 and 311 are treated.

```r
myMethylRawList = methRead(file.list, sample.id = list("301", "302", "310", "311"),
 assembly="NONE", treatment = c(0,0,1,1))
```

{: .success-title }
> STDOUT
>
>Received list of locations.
>
>Reading file.
>
>Reading file.
>
>Reading file.
>
>Reading file.


![alt text](image-66.png)

`myMethylRawList` is an object of the type `MethylRawList`. It is a list of `MethylRaw` objects (S4 class objects for those familiar with R), with as many elements as there are .methylkit files loaded using the list function.
Calling the `MethylRawList` object provides an overview of the loaded methylomes. For each one, the first rows of the table containing the methylation data are shown, along with additional information such as the sample name and the context of the analyzed cytosines.

If we want to access a single element of the list, we can use the following command: `myMethylRawList[[1]]` to access the first element of the list, which is the methylKit object for sample 301. Eitherwise, data can be accessed using `getData()` function, and we have a finer control on the data we want to access. 

Datasets can be manipulated in different ways and for example can be filtered by coverage 

```r
filtered.myMethylRawList = filterByCoverage(myMethylRawList, lo.count=4, 
lo.perc=NULL, hi.count=NULL, hi.perc=99.9)
```


<!--
n questo caso il valore minimo lo imponiamo come valore assoluto (lo.count) e lo settiamo pari a 4 (solitamente si usa 10, ma nel contesto CG possiamo scendere a 4), mentre il valore massimo lo imponiamo come percentile (hi.perc) mantenendo quindi fino al 99.9° percentile (vengono scartate le citosine con coverage più alto corrispondenti allo 0.1% delle citosine totali).
-->

Merge the datasets together

```r
meth = unite(filtered.myMethylRawList, destrand=FALSE, save.db=F)
```

<!--
Con destrand=FALSE indica che non voglio una media delle citosine corrispondenti alla stessa coppia CG (voglio che le due citosine sui 2 filamenti vengano considerate in modo indipendente); imposto anche che non voglio che vengano salvati i dati (save database).
-->

{: .success-title }
> STDOUT
>
> uniting...

The methylBase object should look like
```r
head(meth)
```

![alt text](image-65.png)

<!--
noto che effettivamente quello che abbiamo fatto è stato unire le citosine che presentavano un valore in tutti e 4 i campioni (questa operazione è già stata fatta in un altro contesto, in particolare nella preparazione del tabellone per l’analisi PCA e k-means). Se richiamiamo meth, vediamo il numero di righe totali (citosine):
il numero di citosine non è molto elevato.
-->

If now we would like to get the methylation percentage for each C in common between the four samples

```r
meth_round_percent = round(percMethylation(meth, rowids=T), 2)
```

---

If we want to have the methylation values per window size (e.g. 1000), separately for each input sample, and not the single C positions, tiles could be created by defining the window size and the step size. 

```r
tiles = tileMethylCounts(filtered.myMethylRawList, win.size = 1000, step.size=1000)
```

Now we can merge the information of the tiles of the different samples together. 

```r
tiles_united = unite(tiles, destrand=FALSE, save.db=F)
```

And tiles can be filtered 

```r
filtered.tiles = filterByCoverage(tiles, lo.count=4, lo.perc=NULL, hi.count=NULL, hi.perc=99.9)
```

Output tables can be stored into file using the `write.table` function. 

---
## Differential methylation 
{: .no_toc }

In order to compute the differential methylation between the single Cs or the windows, we can use the `calculateDiffMeth` function. Several parameters are already defined by default. We will only need to set the type of test (chi-squared) and rthe correction type (q-value). We will try with the single C values.

Before we could inspect our dataset. We could perform a PCA 

```r
PCASamples(meth, adj.lim=c(1,1))
```
![alt text](image-68.png)

Or a hierarchical clustering 

```r
clusterSamples(meth, dist="correlation", method="ward", plot=TRUE)
```

![alt text](image-67.png)

Now we are ready to calculate the differential methylation between the two conditions. 

```r
myDiff=calculateDiffMeth(meth)
```

{: .success-title }
> STDOUT
>
> two groups detected:
>
> will calculate methylation difference as the difference of
>
> treatment (group: 1) - control (group: 0)

The created object will have a C in each row, with values of differential methylation. The last column represent the weighted difference of methylation between test and control samples. Positive values will show hypermethylation of the test samples, while negative values will show hypomethylation. 

```r
myDiff
```
![alt text](image-69.png)

In order to get only the significant results

```r
myDiffSignificants=getMethylDiff(myDiff,difference=0,qvalue=0.01)
```

```r
myDiffSignificants
```


For example, C at position 449970, which correspond to row number 22 of our merged dataset, has a differentially methylated value of 44.82292

If we extract the corresponding values from the input table

```r
meth[meth@row.names==22,]
```

We will see that our control samples have low methylation values compared to the treated one, which have methylation values nearly to 1. 

Table can then be saved using the `write.table` function 

```r
write.table(getData(myDiffSignificants),file= "differential_methylation.txt",
 quote=FALSE, sep="\t", row.names=FALSE)
```



The same can be obtained also for the tiles (windows) 

```r
myDiff_tiles=calculateDiffMeth(tiles_united, adjust=c("qvalue"),test=c("Chisq"))
```

By performing the analysis this way we will get the putative windows were differential methylation is observed, which are commonly referred as Differentially Methylated Regions (DMRs). 


The windows could be saved to file and imported in igv to visualise the differentially methylated windows in the genomic context, as for example genes and repetitive elements.